cmd/soroban-cli/src/rpc/transaction.rs

```diff
@@ -0,0 +1,357 @@
+use soroban_env_host::xdr::{
+    ContractAuth, DiagnosticEvent, HostFunction, OperationBody, ReadXdr, SorobanTransactionData,
+    Transaction, TransactionExt, VecM,
+};
+
+use crate::rpc::{Error, LogEvents, SimulateTransactionResponse};
+
+// Apply the result of a simulateTransaction onto a transaction envelope, preparing it for
+// submission to the network.
+pub fn assemble(
+    raw: &Transaction,
+    simulation: &SimulateTransactionResponse,
+    log_events: Option<LogEvents>,
+) -> Result<Transaction, Error> {
+    let mut tx = raw.clone();
+
+    // Right now simulate.results is one-result-per-function, and assumes there is only one
+    // operation in the txn, so we need to enforce that here. I (Paul) think that is a bug
+    // in soroban-rpc.simulateTransaction design, and we should fix it there.
+    // TODO: We should to better handling so non-soroban txns can be a passthrough here.
+    if tx.operations.len() != 1 {
+        return Err(Error::UnexpectedOperationCount {
+            count: tx.operations.len(),
+        });
+    }
+
+    // TODO: Should we keep this?
```

Comment:
The `log_events` piece here is slightly awkward (passing a log fn into this fn). We could instead return the events somehow and do the logging externally where it is needed (only invoke for now). Or we could leave this as-is for now...

Reply:
Given that it has no functional downside, I'd suggest leave it as is for this release and create a ticket to address this for next release.